### PR TITLE
Update and emit events

### LINKED ISSUE #952: Update and emit wallet events
## Description

We need to emit wallet events in all places where needed and add more/change some to the new layouts.

## Motivation

Some users want to do action based on events and for the Ledger Signer it's required that we emit the generated address and show the transaction data in Firefly, so the user can compare it.

## Requirements

Update and emit
1. PreparedTransactionEventData to the new transaction layout (include full outputs)
2. BalanceChangeEvent should be renamed to something like NewOutputEvent and include the output, for more information see https://github.com/iotaledger/wallet.rs/issues/895. It should also be clear if the output can be really owned (no timelock/expiration/storage deposit return unlock condition), maybe for such outputs we should create an own event type.
3. TransactionInclusionEvent (only emit, no need to change the layout)

## Open questions (optional)

Any other events that could be useful?

## Are you planning to do it yourself in a pull request?

No.

### PR DESCRIPTION
**Pull request by Thoralf-M**
# Description of change

Update and emit events and update ledger address generation for it

## Links to any relevant issues

Fixes #952 

## Type of change

- Breaking change (fix or feature that would cause existing functionality to not work as expected)

## How the change has been tested

Describe the tests that you ran to verify your changes.

Make sure to provide instructions for the maintainer as well as any relevant configurations.

## Change checklist

- [ ] I have followed the contribution guidelines for this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked that new and existing unit tests pass locally with my changes

### CODE REVIEW

File: src/events/types.rs [hunk 1]
```diff
@@ -16,64 +19,65 @@ pub struct Event {
     pub event: WalletEvent,
 }
 
-// do we want an event for transaction confirmation or if it failed?
-// event for new detected outputs?
-
-#[derive(Clone, Debug, Eq, PartialEq, Hash, Serialize, Deserialize)]
+#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
 pub enum WalletEvent {
-    BalanceChange(BalanceChangeEvent),
```
Comment (thibault-martinez) [inline src/events/types.rs:24]: Can you order them the same way as WalletEventType ?

File: src/account/operations/syncing/mod.rs [hunk 2]
```diff
@@ -206,10 +223,25 @@ impl AccountHandle {
         }
 
         // Add new synced outputs
-        for output in new_outputs {
-            account.outputs.insert(output.output_id, output.clone());
-            if !output.is_spent {
-                account.unspent_outputs.insert(output.output_id, output);
+        for output_data in unspent_outputs {
+            // Inser output, if it's unknown emit the NewOutputEvent
```
Comment (thibault-martinez) [inline src/account/operations/syncing/mod.rs]:
```suggestion
            // Insert output, if it's unknown emit the NewOutputEvent
```

File: src/account/operations/address_generation.rs [hunk 3]
```diff
@@ -83,58 +83,91 @@ impl AccountHandle {
...
+                            .read()
+                            .await
+                            .generate_addresses(
+                                account.coin_type,
+                                account.index,
+                                address_index..address_index + 1,
+                                options.internal,
+                                options.metadata.clone(),
+                            )
+                            .await?;
```
Comment (thibault-martinez) [inline src/account/operations/address_generation.rs:139]: What's different between the two generations ?

Comment (Thoralf-M) [reply src/account/operations/address_generation.rs:139]: The first one will just generate the address and return it direclty, the second one shows the address on the screen of the ledger nano and requires physical approval to continue

Comment (thibault-martinez) [reply src/account/operations/address_generation.rs:139]: Sorry if I'm being stupid but the two generation are exactly the same so why can't we just do the emit under the feature ?

Comment (Thoralf-M) [reply src/account/operations/address_generation.rs:139]: The first one has `changed_metadata.clone(),`  where syncing is set to true, which makes the ledger to not display the address and return without user approval
Then the event with the address gets emitted and Firefly can display the address, so when the second generation is called , with `options.metadata.clone()` and the ledger displays the address, the user can compare it with the address displayed in Firefly

Comment (thibault-martinez) [reply src/account/operations/address_generation.rs:139]: Ah damn I don't know how I missed changed_metadata != options.metadata, sorry about that 🤦🏻

**Review by thibault-martinez — APPROVED** [review]

**Review by Tuditi — APPROVED** [review]
LGTM!

**Pull request merged**